ra/ra.go

```diff
@@ -995,60 +996,86 @@ func (ra *RegistrationAuthorityImpl) FinalizeOrder(ctx context.Context, req *rap
 		return nil, err
 	}
 
-	// Step 3: Issue the Certificate
-	cert, err := ra.issueCertificateInner(
-		ctx, csr, accountID(req.Order.RegistrationID), orderID(req.Order.Id))
-
-	// Step 4: Fail the order if necessary, and update metrics and log fields
-	var result string
+	// Update the order status locally since the SA doesn't return the updated
+	// order itself after setting the status
 	order := req.Order
-	if err != nil {
-		// The problem is computed using `web.ProblemDetailsForError`, the same
-		// function the WFE uses to convert between `berrors` and problems. This
-		// will turn normal expected berrors like berrors.UnauthorizedError into the
-		// correct `urn:ietf:params:acme:error:unauthorized` problem while not
-		// letting anything like a server internal error through with sensitive
-		// info.
-		ra.failOrder(ctx, req.Order, web.ProblemDetailsForError(err, "Error finalizing order"))
-
-		// Update the order status locally since the SA doesn't return the updated
-		// order itself after setting the status
-		order.Status = string(core.StatusInvalid)
-
-		logEvent.Error = err.Error()
-		beeline.AddFieldToTrace(ctx, "issuance.error", err)
-		result = "error"
-	} else {
-		// Update the order status locally since the SA doesn't return the updated
-		// order itself after setting the status
-		order.CertificateSerial = core.SerialToString(cert.SerialNumber)
-		order.Status = string(core.StatusValid)
+	order.Status = string(core.StatusInvalid)
+
+	// We do steps 3 and 4 in a goroutine so that we can better handle latency
+	// from getting SCTs and writing the (pre)certificate to the database. This
+	// lets us return the order in the Processing state to the client immediately,
+	// prompting them to poll the Order object and wait for it to be put into its
+	// final state.
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		// Step 3: Issue the Certificate
+		if features.Enabled(features.AsyncFinalize) {
+			// If we're in async mode, use a context with a much longer timeout.
+			// TODO(go1.22?): use context.Detach to preserve tracing metadata.
+			var cancel func()
+			ctx, cancel = context.WithTimeout(context.Background(), 5*time.Minute)
```

Comment:
This has some surprising scope implications since it's setting ctx in the parent scope too. I think it would be better to do the "if async, detach context" work inside `issuerCertificateInner`, since we expect that whole function's ctx to be affected.

Also, the timeout for finalization contexts should be configurable.

Reply:
Good calls, both done.